### PR TITLE
Attempt to fix the webauthn migration again - part 3

### PR DESCRIPTION
**Pull request by zeripath**
v208.go is seriously broken as it misses an ID() check. We need to no-op and remigrate all of the u2f keys.

See #18756

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

File: models/auth/webauthn.go [hunk 1]
```diff
@@ -43,7 +43,7 @@ type WebAuthnCredential struct {
 	Name            string
 	LowerName       string `xorm:"unique(s)"`
 	UserID          int64  `xorm:"INDEX unique(s)"`
-	CredentialID    string `xorm:"INDEX"`
+	CredentialID    string `xorm:"INDEX VARCHAR(500)"`
```
Comment (fnetX) [inline models/auth/webauthn.go]: This should no longer be necessary? Same obviously applies to the backport + all other places that are touched in this PR? All calculations done were probably correct, MariaDB was probably right this time and the comments about MariaDB probably just lead to confusion in the future? :wink:

Comment (zeripath) [reply models/auth/webauthn.go]: If you want to show that 410 works then fine: We can change it back.

File: models/migrations/v209.go [hunk 2]
```diff
@@ -5,140 +5,13 @@
...
-		}
-
-		if len(regs) < 50 {
-			break
-		}
-		start += 50
-		regs = regs[:0]
-	}
+	// no-op
+	// MariaDB asserts that 408 characters is too long for a VARCHAR(410) so this migration is clearly wrong.
```
Comment (zeripath) [inline models/migrations/v209.go]:
```suggestion
	// v208 was completely wrong
```

File: models/auth/webauthn.go (hunk 1 above)
Comment (fnetX) [reply models/auth/webauthn.go]: Migrated with https://github.com/go-gitea/gitea/pull/18756#issuecomment-1039711413:

~~~
select max(octet_length(credential_id)) from webauthn_credential;
+----------------------------------+
| max(octet_length(credential_id)) |
+----------------------------------+
|                              408 |
+----------------------------------+
1 row in set (0.000 sec)
~~~

the final issue that caused the length was at least unrelated to the column size. And it's plausible after all, so I think we can go back to the lower limit. Or at least remove the confusing comments and just say we have some extra space ...

Still, we wouldn't have discovered this issue if it was 500, as it's kinda a time bomb. It just silently breaks for small instances with ~ 450 credentials when it is about 500. The expected size should be 408, if it is more, we should see an error and investigate. Same in the future.

**Comment by zeripath** [comment]
I think this might actually still be insufficient. 

Is it possible that v208 might have updated the IDs to all be the same number too?

If that's the case then this won't remigrate.

Then there's the case of the users who have deleted the old u2f keys already too.

File: models/migrations/v210.go [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+			c := &webauthnCredential{
+				ID:              reg.ID,
+				Name:            reg.Name,
+				LowerName:       strings.ToLower(reg.Name),
+				UserID:          reg.UserID,
+				CredentialID:    base32.HexEncoding.EncodeToString(parsed.KeyHandle),
+				PublicKey:       elliptic.Marshal(elliptic.P256(), parsed.PubKey.X, parsed.PubKey.Y),
+				AttestationType: "fido-u2f",
+				AAGUID:          []byte{},
+				SignCount:       reg.Counter,
```
Comment (lunny) [inline models/migrations/v210.go]: Please also copy created and updated and use `x.ID(c.ID).AllCols().Update(c)` to ensure all columns should be updated.

File: models/migrations/v210.go [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+				UpdatedUnix:     reg.UpdatedUnix,
+				CreatedUnix:     reg.CreatedUnix,
+			}
+
+			has, err := x.ID(reg.ID).Where("id = ?", reg.ID).Get(new(webauthnCredential))
+			if err != nil {
+				return fmt.Errorf("unable to get webauthn_credential[%d]. Error: %v", reg.ID, err)
+			}
+			if !has {
+				_, err = x.ID(remigrated.ID).AllCols().Insert(remigrated)
```
Comment (lunny) [inline models/migrations/v210.go]: just `x.Insert`

Comment (zeripath) [reply models/migrations/v210.go]:
```suggestion
				_, err = x.Insert(remigrated)
```

**Review by lunny — APPROVED** [review]

**Review by fnetX — APPROVED** [review]

**Comment by zeripath** [comment]
OK I think this is finally ready

**Review by Gusted — APPROVED** [review]
Let's hope it be the last one 😄 Couldn't find any issue with it.

**Pull request merged**